File: Reference/src/Brontide.Reference.Experimental.Binding/Portable/PortableNegotiation.cs
using System.Collections.Immutable;

namespace Brontide.Reference.Experimental.Binding.Portable;

/// <summary>
/// Negotiates two contract documents into one frozen Binding Plan.
/// </summary>
/// <remarks>
/// Every step completes before any provider activation, so a failure at any step produces zero
/// provider effects. Version 0.1 negotiates by exact reference equality: there is no structural
/// similarity matching, no nearest-version selection, and no inference of authority from provider
/// availability.
/// </remarks>
public static class PortableNegotiation
{
    private static readonly ImmutableArray<string> SupportedFeatures =
    [
        "establishment", "readiness-signal", "single-invocation", "clean-withdrawal", "clean-termination"
    ];

    private static readonly ImmutableArray<string> UnsupportedFeatures =
    [
        "retry", "cancellation", "streaming", "ordering-guarantee", "exactly-once-execution"
    ];

    public static PortableBindingPlan Negotiate(
        PortableContractDocument required,
        PortableContractDocument offered,
        PortableRealization realization,
        string hostEndpoint,
        string providerEndpoint,
        string selectionReason)
    {
        ArgumentNullException.ThrowIfNull(required);
        ArgumentNullException.ThrowIfNull(offered);

        RequireContractVersion(required, offered);
        RequireProvider(required, offered);
        RequireComponent(required, offered);
        var satisfied = MatchRequirements(required, offered);
        MatchOperations(required, offered);
        MatchAuthority(required, offered);
        MatchRepresentation(required, offered);
        MatchLimitsAndLifecycle(required, offered);

        // Compact identifiers are assigned only once every canonical step above has succeeded.
        var operations = required.Operations.Select(operation => operation.Reference).ToImmutableArray();
        var shapes = required.Shapes.Select(shape => shape.Reference).Order().ToImmutableArray();
        var fragments = required.Fragments.Select(fragment => fragment.Reference).Order().ToImmutableArray();

        return new PortableBindingPlan(
            PortablePlanId.New(),
            required,
            offered.Provider,
            operations,
            shapes,
            fragments,
            satisfied,
            AssignCompactIdentifiers(operations, shapes, fragments),
            hostEndpoint,
            providerEndpoint,
            selectionReason,
            realization);
    }

    private static void RequireContractVersion(PortableContractDocument required, PortableContractDocument offered)
    {
        if (required.ContractVersion != offered.ContractVersion ||
            required.ContractVersion != PortableContractDocument.SupportedContractVersion)
        {
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedVersion,
                "contract-version",
                "The declared contract versions are not both the recognized version.");
        }
    }

    /// <summary>
    /// A required document naming a provider is binding, not expectational.
    /// </summary>
    /// <remarks>
    /// Version 0.1 defines no way to say "any provider of this Component", so the peer must answer as
    /// the provider the host named. Refusing the mismatch here is what lets the plan's provider facts
    /// be read from the offered document and still equal the requirement. Decision 11.
    /// </remarks>
    private static void RequireProvider(PortableContractDocument required, PortableContractDocument offered)
    {
        if (required.Provider != offered.Provider)
        {
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                "provider-mismatch",
                $"Required provider {required.Provider} answered as {offered.Provider}.");
        }
    }

    private static void RequireComponent(PortableContractDocument required, PortableContractDocument offered)
    {
        if (required.Component != offered.Component)
        {
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                "component-mismatch",
                $"Required component {required.Component} was not offered as the exact contract.");
        }
    }

    private static ImmutableArray<PortableDependencyReference> MatchRequirements(
        PortableContractDocument required,
        PortableContractDocument offered)
    {
        var satisfied = ImmutableArray.CreateBuilder<PortableDependencyReference>();
        foreach (var requirement in required.Requirements)
        {
            var provided = offered.Provisions.Any(provision =>
                provision.Kind == requirement.Kind &&
                provision.Reference == requirement.Reference &&
                provision.ProviderSpecific == requirement.ProviderSpecific);

            switch (requirement.Strength)
            {
                case PortableRequirementStrength.Required when !provided:
                    throw new PortableFaultException(
                        PortableProtocolCategory.UnsupportedContract,
                        "requirement-unmet",
                        $"Required {requirement.Kind.Token()} {requirement.Reference} was not offered.");
                case PortableRequirementStrength.Opposed when provided:
                    throw new PortableFaultException(
                        PortableProtocolCategory.UnsupportedContract,
                        "opposed-requirement-offered",
                        $"Opposed {requirement.Kind.Token()} {requirement.Reference} was offered; it is refused rather than ignored.");
                default:
                    break;
            }

            if (provided && requirement.Strength != PortableRequirementStrength.Opposed)
            {
                satisfied.Add(requirement.Reference);
            }
        }

        return satisfied.ToImmutable();
    }

    private static void MatchOperations(PortableContractDocument required, PortableContractDocument offered)
    {
        foreach (var operation in required.Operations)
        {
            var candidate = offered.Operations
                .FirstOrDefault(offeredOperation => offeredOperation.Reference == operation.Reference)
                ?? throw new PortableFaultException(
                    PortableProtocolCategory.UnsupportedOperation,
                    "unknown-operation",
                    $"Operation {operation.Reference} is not offered.");

            if (candidate.InputShape != operation.InputShape ||
                candidate.ResultShape != operation.ResultShape ||
                candidate.DetailShape != operation.DetailShape ||
                !candidate.RequiredFragments.SequenceEqual(operation.RequiredFragments) ||
                !candidate.ResourceFlavors.SequenceEqual(operation.ResourceFlavors, StringComparer.Ordinal))
            {
                throw new PortableFaultException(
                    PortableProtocolCategory.UnsupportedContract,
                    "operation-shape-mismatch",
                    $"Operation {operation.Reference} declares incompatible Shapes, Fragments, or resource flavors.");
            }

            foreach (var flavor in operation.ResourceFlavors)
            {
                _ = PortableResourceFlavors.Parse(flavor);
            }
        }
    }

    private static void MatchAuthority(PortableContractDocument required, PortableContractDocument offered)
    {
        PortableAuthorityGate.RequireValidPresentation(required.Authority);
        PortableAuthorityGate.RequireValidPresentation(offered.Authority);
        if (required.Authority != offered.Authority)
        {
            throw new PortableFaultException(
                PortableProtocolCategory.InvalidAuthorityPresentation,
                "authority-mismatch",
                "The endpoints declare different authority presentations, which cannot be resolved to a permissive mode.");
        }
    }

    private static void MatchRepresentation(PortableContractDocument required, PortableContractDocument offered)
    {
        // Compared member by member: the declaration carries sequences, whose record equality is by
        // instance rather than by content.
        var equal = string.Equals(
                required.Representation.Representation,
                offered.Representation.Representation,
                StringComparison.Ordinal) &&
            string.Equals(required.Representation.Framing, offered.Representation.Framing, StringComparison.Ordinal) &&
            required.Representation.ResourceFlavors.SequenceEqual(
                offered.Representation.ResourceFlavors,
                StringComparer.Ordinal) &&
            required.Representation.AcceptedResourceHandles.SequenceEqual(
                offered.Representation.AcceptedResourceHandles,
                StringComparer.Ordinal);
        if (!equal)
        {
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                "representation-mismatch",
                "The endpoints declare different payload representations.");
        }

        if (!string.Equals(
                required.Representation.Representation,
                PortableRepresentations.PortableCborCore,
                StringComparison.Ordinal))
        {
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                "representation-unsupported",
                $"'{required.Representation.Representation}' is not the portable wire representation.");
        }

        if (!string.Equals(
                required.Representation.Framing,
                PortableRepresentations.LengthDelimited,
                StringComparison.Ordinal))
        {
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                "framing-unsupported",
                "The portable representation is length-delimited.");
        }

        foreach (var flavor in required.Representation.ResourceFlavors)
        {
            _ = PortableResourceFlavors.Parse(flavor);
        }

        if (required.Representation.AcceptedResourceHandles.Length > 0 &&
            !required.Representation.ResourceFlavors.Contains(
                PortableResourceFlavors.AddressingOnlyHandle,
                StringComparer.Ordinal))
        {
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                "handle-accept-list",
                "An accept list is declared without the addressing-only-handle flavor it governs.");
        }
    }

    private static void MatchLimitsAndLifecycle(PortableContractDocument required, PortableContractDocument offered)
    {
        if (required.Limits != offered.Limits)
        {
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                "limits-mismatch",
                "The endpoints declare different delivery limits.");
        }

        required.Limits.Validate();

        if (required.Lifecycle.ReplayProtectionDeclared != offered.Lifecycle.ReplayProtectionDeclared ||
            !string.Equals(required.Lifecycle.ReplayWindow, offered.Lifecycle.ReplayWindow, StringComparison.Ordinal))
        {
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                "replay-declaration-mismatch",
                "The endpoints declare different replay protection.");
        }

        foreach (var feature in SupportedFeatures)
        {
            RequireFeature(required, feature, expected: true);
            RequireFeature(offered, feature, expected: true);
        }

        foreach (var feature in UnsupportedFeatures)
        {
            RequireFeature(required, feature, expected: false);
            RequireFeature(offered, feature, expected: false);
        }
    }

    private static void RequireFeature(PortableContractDocument document, string feature, bool expected)
    {
        if (!document.Lifecycle.Features.TryGetValue(feature, out var declared))
        {
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                "feature-undeclared",
                $"Lifecycle feature '{feature}' is not declared, so no consumer can tell whether it is promised.");
        }

        if (declared != expected)
        {
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                expected ? "feature-unsupported" : "feature-not-promised",
                expected
                    ? $"Lifecycle feature '{feature}' is required and must be supported."
                    : $"Lifecycle feature '{feature}' is not promised by version 0.1 and is refused rather than silently ignored.");
        }
    }

    private static ImmutableArray<PortableCompactAssignment> AssignCompactIdentifiers(
        ImmutableArray<PortableOperationReference> operations,
        ImmutableArray<PortableShapeReference> shapes,
        ImmutableArray<PortableFragmentReference> fragments)
    {
        var assignments = ImmutableArray.CreateBuilder<PortableCompactAssignment>();
        for (var index = 0; index < operations.Length; index++)
        {
            assignments.Add(new PortableCompactAssignment(
                PortableIdentitySpace.Operation,
                operations[index].ToString(),
                new PortableCompactIdentifier(index)));
        }

        for (var index = 0; index < shapes.Length; index++)
        {
            assignments.Add(new PortableCompactAssignment(
                PortableIdentitySpace.Shape,
                shapes[index].ToString(),
                new PortableCompactIdentifier(index)));
        }

        for (var index = 0; index < fragments.Length; index++)
        {
            assignments.Add(new PortableCompactAssignment(
                PortableIdentitySpace.Fragment,
                fragments[index].ToString(),
                new PortableCompactIdentifier(index)));
        }

        return assignments.ToImmutable();
    }
}

File: Minimal/src/Brontide.Minimal.Binding/Portable/PortableNegotiation.fs
namespace Brontide.Minimal.Binding.Portable

/// Negotiates two contract documents into one frozen Binding Plan.
///
/// Every step completes before any provider activation, so a failure at any step produces zero
/// provider effects. Version 0.1 negotiates by exact reference equality: there is no structural
/// similarity matching, no nearest-version selection, and no inference of authority from provider
/// availability.
[<RequireQualifiedAccess>]
module PortableNegotiation =

    let private supportedFeatures =
        [ "establishment"; "readiness-signal"; "single-invocation"; "clean-withdrawal"; "clean-termination" ]

    let private unsupportedFeatures =
        [ "retry"; "cancellation"; "streaming"; "ordering-guarantee"; "exactly-once-execution" ]

    let private requireContractVersion (required: ContractDocument) (offered: ContractDocument) =
        ensure
            (required.ContractVersion = offered.ContractVersion
             && required.ContractVersion = ContractDocument.SupportedContractVersion)
            (fun () ->
                refuse
                    ProtocolCategory.UnsupportedVersion
                    "contract-version"
                    "The declared contract versions are not both the recognized version.")

    /// A required document naming a provider is binding, not expectational.
    ///
    /// Version 0.1 defines no way to say "any provider of this Component", so the peer must answer as
    /// the provider the host named. Refusing the mismatch here is what lets the plan's provider facts
    /// be read from the offered document and still equal the requirement. Decision 11.
    let private requireProvider (required: ContractDocument) (offered: ContractDocument) =
        ensure (required.Provider = offered.Provider) (fun () ->
            unsupportedContract
                "provider-mismatch"
                $"Required provider {PortableProviderRef.text required.Provider} answered as {PortableProviderRef.text offered.Provider}.")

    let private requireComponent (required: ContractDocument) (offered: ContractDocument) =
        ensure (required.Component = offered.Component) (fun () ->
            unsupportedContract
                "component-mismatch"
                $"Required component {PortableComponentRef.text required.Component} was not offered as the exact contract.")

    let private matchRequirements (required: ContractDocument) (offered: ContractDocument) =
        let isProvided (requirement: Requirement) =
            offered.Provisions
            |> List.exists (fun provision ->
                provision.Kind = requirement.Kind
                && provision.Reference = requirement.Reference
                && provision.ProviderSpecific = requirement.ProviderSpecific)

        let rec walk satisfied remaining =
            match remaining with
            | [] -> Ok(List.rev satisfied)
            | (requirement: Requirement) :: tail ->
                let provided = isProvided requirement

                match requirement.Strength with
                | RequirementStrength.Required when not provided ->
                    unsupportedContract
                        "requirement-unmet"
                        $"Required {DependencyKind.token requirement.Kind} {PortableDependencyRef.text requirement.Reference} was not offered."
                | RequirementStrength.Opposed when provided ->
                    unsupportedContract
                        "opposed-requirement-offered"
                        $"Opposed {DependencyKind.token requirement.Kind} {PortableDependencyRef.text requirement.Reference} was offered; it is refused rather than ignored."
                | RequirementStrength.Opposed -> walk satisfied tail
                | RequirementStrength.Required
                | RequirementStrength.Preferred
                | RequirementStrength.Optional ->
                    walk (if provided then requirement.Reference :: satisfied else satisfied) tail

        walk [] required.Requirements

    let private matchOperations (required: ContractDocument) (offered: ContractDocument) =
        required.Operations
        |> iterate (fun operation ->
            match ContractDocument.tryOperation operation.Reference offered with
            | None ->
                refuse
                    ProtocolCategory.UnsupportedOperation
                    "unknown-operation"
                    $"Operation {PortableOperationRef.text operation.Reference} is not offered."
            | Some candidate ->
                portable {
                    do!
                        ensure (candidate = operation) (fun () ->
                            unsupportedContract
                                "operation-shape-mismatch"
                                $"Operation {PortableOperationRef.text operation.Reference} declares incompatible Shapes, Fragments, or resource flavors.")

                    do! operation.ResourceFlavors |> iterate (ResourceFlavor.tryParse >> Result.map ignore)
                })

    let private matchAuthority (required: ContractDocument) (offered: ContractDocument) =
        portable {
            do! PortableAuthorityGate.requireValidPresentation required.Authority
            do! PortableAuthorityGate.requireValidPresentation offered.Authority

            do!
                ensure (required.Authority = offered.Authority) (fun () ->
                    invalidAuthority
                        "authority-mismatch"
                        "The endpoints declare different authority presentations, which cannot be resolved to a permissive mode.")
        }

    let private matchRepresentation (required: ContractDocument) (offered: ContractDocument) =
        portable {
            do!
                ensure (required.Representation = offered.Representation) (fun () ->
                    unsupportedContract "representation-mismatch" "The endpoints declare different payload representations.")

            do!
                ensure (required.Representation.Representation = PortableRepresentations.PortableCborCore) (fun () ->
                    unsupportedContract
                        "representation-unsupported"
                        $"'{required.Representation.Representation}' is not the portable wire representation.")

            do!
                ensure (required.Representation.Framing = PortableRepresentations.LengthDelimited) (fun () ->
                    unsupportedContract "framing-unsupported" "The portable representation is length-delimited.")

            do! required.Representation.ResourceFlavors |> iterate (ResourceFlavor.tryParse >> Result.map ignore)

            do!
                ensure
                    (List.isEmpty required.Representation.AcceptedResourceHandles
                     || List.contains ResourceFlavor.AddressingOnlyHandleToken required.Representation.ResourceFlavors)
                    (fun () ->
                        unsupportedContract
                            "handle-accept-list"
                            "An accept list is declared without the addressing-only-handle flavor it governs.")
        }

    let private requireFeature (document: ContractDocument) feature expected =
        match Map.tryFind feature document.Lifecycle.Features with
        | None ->
            unsupportedContract
                "feature-undeclared"
                $"Lifecycle feature '{feature}' is not declared, so no consumer can tell whether it is promised."
        | Some declared when declared <> expected ->
            if expected then
                unsupportedContract "feature-unsupported" $"Lifecycle feature '{feature}' is required and must be supported."
            else
                unsupportedContract
                    "feature-not-promised"
                    $"Lifecycle feature '{feature}' is not promised by version 0.1 and is refused rather than silently ignored."
        | Some _ -> Ok()

    let private matchLimitsAndLifecycle (required: ContractDocument) (offered: ContractDocument) =
        portable {
            do!
                ensure (required.Limits = offered.Limits) (fun () ->
                    unsupportedContract "limits-mismatch" "The endpoints declare different delivery limits.")

            do! PortableLimits.validate required.Limits

            do!
                ensure
                    (required.Lifecycle.ReplayProtectionDeclared = offered.Lifecycle.ReplayProtectionDeclared
                     && required.Lifecycle.ReplayWindow = offered.Lifecycle.ReplayWindow)
                    (fun () ->
                        unsupportedContract "replay-declaration-mismatch" "The endpoints declare different replay protection.")

            for feature in supportedFeatures do
                do! requireFeature required feature true
                do! requireFeature offered feature true

            for feature in unsupportedFeatures do
                do! requireFeature required feature false
                do! requireFeature offered feature false
        }

    /// Compact identifiers are assigned only once every canonical step has succeeded, so a compact
    /// identifier can never precede the negotiation that gives it meaning.
    let private assignCompactIdentifiers operations shapes fragments =
        let assign space text index =
            CompactId.tryCreate index
            |> Result.map (fun compact ->
                { Space = space
                  Reference = text
                  Compact = compact })

        portable {
            let! operationAssignments =
                operations
                |> List.mapi (fun index reference -> IdentitySpace.Operation, PortableOperationRef.text reference, index)
                |> traverse (fun (space, text, index) -> assign space text index)

            let! shapeAssignments =
                shapes
                |> List.mapi (fun index reference -> IdentitySpace.Shape, PortableShapeRef.text reference, index)
                |> traverse (fun (space, text, index) -> assign space text index)

            let! fragmentAssignments =
                fragments
                |> List.mapi (fun index reference -> IdentitySpace.Fragment, PortableFragmentRef.text reference, index)
                |> traverse (fun (space, text, index) -> assign space text index)

            return operationAssignments @ shapeAssignments @ fragmentAssignments
        }

    let negotiate
        (required: ContractDocument)
        (offered: ContractDocument)
        (realization: Realization)
        (hostEndpoint: string)
        (providerEndpoint: string)
        (selectionReason: string)
        : PortableResult<BindingPlan> =
        portable {
            do! requireContractVersion required offered
            do! requireProvider required offered
            do! requireComponent required offered
            let! satisfied = matchRequirements required offered
            do! matchOperations required offered
            do! matchAuthority required offered
            do! matchRepresentation required offered
            do! matchLimitsAndLifecycle required offered

            let! catalog = ShapeCatalog.fromContract required
            let operations = required.Operations |> List.map (fun operation -> operation.Reference)
            let shapes = required.Shapes |> List.map (fun shape -> shape.Reference) |> List.sort
            let fragments = required.Fragments |> List.map (fun fragment -> fragment.Reference) |> List.sort
            let! assignments = assignCompactIdentifiers operations shapes fragments

            return
                BindingPlan.freeze
                    (PlanId.next ())
                    required
                    offered.Provider
                    catalog
                    operations
                    shapes
                    fragments
                    satisfied
                    assignments
                    hostEndpoint
                    providerEndpoint
                    selectionReason
                    realization
        }
